From f55cc58325ed5df0b0d5fd0a61ebb32d735c2481 Mon Sep 17 00:00:00 2001
From: Haihua Hu <jared.hu@nxp.com>
Date: Mon, 7 Nov 2016 16:31:02 +0800
Subject: [PATCH 23/24] qmlglsrc: some enhancements for qmlglsrc

1. Need set use-default-fbo to qquickwindow during set
   property to support change render target on the fly.
2. Calculate qmlglsrc refresh frame rate in qtglwindow

Upstream-Status: Backport[1.11.1]

https://bugzilla.gnome.org/show_bug.cgi?id=774035

Signed-off-by: Haihua Hu <jared.hu@nxp.com>
---
 ext/qt/gstqtsrc.cc | 14 +++++---------
 ext/qt/qtwindow.cc | 34 ++++++++++++++++++++--------------
 ext/qt/qtwindow.h  |  1 -
 3 files changed, 25 insertions(+), 24 deletions(-)

diff --git a/ext/qt/gstqtsrc.cc b/ext/qt/gstqtsrc.cc
index 3cfc709..350f70d 100644
--- a/ext/qt/gstqtsrc.cc
+++ b/ext/qt/gstqtsrc.cc
@@ -132,6 +132,7 @@ gst_qt_src_init (GstQtSrc * src)
   gst_base_src_set_live (GST_BASE_SRC (src), DEFAULT_IS_LIVE);
   src->default_fbo = FALSE;
   src->pending_image_orientation = TRUE;
+  src->window = NULL;
 }
 
 static void
@@ -147,6 +148,8 @@ gst_qt_src_set_property (GObject * object, guint prop_id,
 
       if (qt_src->window)
         delete qt_src->window;
+      qt_src->window = NULL;
+
       if (qt_src->qwindow)
         qt_src->window = new QtGLWindow (NULL, qt_src->qwindow);
 
@@ -154,6 +157,8 @@ gst_qt_src_set_property (GObject * object, guint prop_id,
     }
     case PROP_DEFAULT_FBO:
       qt_src->default_fbo = g_value_get_boolean (value);
+      if (qt_src->window)
+        qt_window_use_default_fbo (qt_src->window, qt_src->default_fbo);
       break;
     default:
       G_OBJECT_WARN_INVALID_PROPERTY_ID (object, prop_id, pspec);
@@ -508,7 +513,6 @@ gst_qt_src_change_state (GstElement * element, GstStateChange transition)
   GstQtSrc *qt_src = GST_QT_SRC (element);
   GstStateChangeReturn ret = GST_STATE_CHANGE_SUCCESS;
   QGuiApplication *app;
-  guint64 frames_rendered = 0;
 
   GST_DEBUG ("changing state: %s => %s",
       gst_element_state_get_name (GST_STATE_TRANSITION_CURRENT (transition)),
@@ -553,18 +557,10 @@ gst_qt_src_change_state (GstElement * element, GstStateChange transition)
 
   switch (transition) {
     case GST_STATE_CHANGE_PLAYING_TO_PAUSED:
-      qt_src->run_time = gst_element_get_start_time (GST_ELEMENT (qt_src));
       break;
     case GST_STATE_CHANGE_PAUSED_TO_READY:
       break;
     case GST_STATE_CHANGE_READY_TO_NULL:
-      qt_window_get_total_frames (qt_src->window, &frames_rendered);
-      if (qt_src->run_time > 0) {
-        GST_DEBUG ("qmlglsrc Total refresh frames (%ld), playing for (%"
-            GST_TIME_FORMAT "), fps (%.3f).\n", frames_rendered,
-            GST_TIME_ARGS (qt_src->run_time),
-            (gfloat) GST_SECOND * frames_rendered / qt_src->run_time);
-      }
       break;
     default:
       break;
diff --git a/ext/qt/qtwindow.cc b/ext/qt/qtwindow.cc
index 3d96751..edb6b24 100644
--- a/ext/qt/qtwindow.cc
+++ b/ext/qt/qtwindow.cc
@@ -29,6 +29,7 @@
 #include "gstqsgtexture.h"
 #include "gstqtglutility.h"
 
+#include <QtCore/QDateTime>
 #include <QtCore/QRunnable>
 #include <QtGui/QGuiApplication>
 #include <QtQuick/QQuickWindow>
@@ -71,6 +72,8 @@ struct _QtGLWindowPrivate
 
   /* frames that qmlview rendered in its gl thread */
   guint64 frames_rendered;
+  qint64 start;
+  qint64 stop;
 };
 
 class InitQtGLContext : public QRunnable
@@ -146,6 +149,12 @@ QtGLWindow::beforeRendering()
 
   g_mutex_lock (&this->priv->lock);
 
+  static volatile gsize once = 0;
+  if (g_once_init_enter(&once)) {
+    this->priv->start = QDateTime::currentDateTime().toMSecsSinceEpoch();
+    g_once_init_leave(&once,1);
+  }
+
   if (!fbo && !this->priv->useDefaultFbo) {
 
     width = source->width();
@@ -157,6 +166,10 @@ QtGLWindow::beforeRendering()
           QOpenGLFramebufferObject::NoAttachment, GL_TEXTURE_2D, GL_RGBA));
 
     source->setRenderTarget(fbo.data());
+  } else if (this->priv->useDefaultFbo) {
+    GST_DEBUG ("use default fbo for renderTargetId");
+    fbo.reset(NULL);
+    source->setRenderTarget(NULL);
   }
 
   g_mutex_unlock (&this->priv->lock);
@@ -256,9 +269,14 @@ QtGLWindow::aboutToQuit()
   this->priv->quit = TRUE;
   g_cond_signal (&this->priv->update_cond);
 
-  g_mutex_unlock (&this->priv->lock);
+  this->priv->stop = QDateTime::currentDateTime().toMSecsSinceEpoch();
+  qint64 duration = this->priv->stop - this->priv->start;
+  float fps = ((float)this->priv->frames_rendered / duration * 1000);
+
+  GST_DEBUG("about to quit, total refresh frames (%lld) in (%0.3f) seconds, fps: %0.3f",
+      this->priv->frames_rendered, (float)duration / 1000, fps);
 
-  GST_DEBUG("about to quit");
+  g_mutex_unlock (&this->priv->lock);
 }
 
 void
@@ -410,15 +428,3 @@ qt_window_use_default_fbo (QtGLWindow * qt_window, gboolean useDefaultFbo)
 
   g_mutex_unlock (&qt_window->priv->lock);
 }
-
-void
-qt_window_get_total_frames (QtGLWindow * qt_window, guint64 *frames)
-{
-  g_return_if_fail (qt_window != NULL);
-
-  g_mutex_lock (&qt_window->priv->lock);
-
-  *frames = qt_window->priv->frames_rendered;
-
-  g_mutex_unlock (&qt_window->priv->lock);
-}
diff --git a/ext/qt/qtwindow.h b/ext/qt/qtwindow.h
index 8cb5a2c..697f389 100644
--- a/ext/qt/qtwindow.h
+++ b/ext/qt/qtwindow.h
@@ -65,7 +65,6 @@ GstGLContext *  qt_window_get_qt_context (QtGLWindow * qt_window);
 GstGLDisplay *  qt_window_get_display (QtGLWindow * qt_window);
 gboolean        qt_window_is_scenegraph_initialized (QtGLWindow * qt_window);
 void            qt_window_use_default_fbo (QtGLWindow * qt_window, gboolean useDefaultFbo);
-void            qt_window_get_total_frames (QtGLWindow * qt_window, guint64 *frames);
 }
 
 #endif /* __QT_WINDOW_H__ */
-- 
1.9.1

